### PR TITLE
Local storage namespace uses appCode and username.

### PR DESCRIPTION
**Pull request by TomTirapani**
See issue #674

### CODE REVIEW

**Comment by amcclain** [comment]
Thanks for grabbing this.  We need to decide if/when to merge as this will be a breaking change and reset any current local state, which currently includes grid state...

Some things to consider:
* Most real users will be the only ones logging in to a site under their overall OS profile. However for apps like Toolbox where you can log out / log in interactively - and also importantly for impersonation - this closes a very real gap in how we were namespacing local storage.
* I can see the value of hashing the appCode and username to not have usernames embedded directly into local storage keys, although I don't know if we have any real-world concerns here given that everything is under an OS profile. The tradeoff is that this makes troubleshooting potentially more difficult.

Would it be overkill to consider a migration strategy, where LocalStorageService would check on init if there was a local namespace with the appName and, if so, migrate its contents to the new desired namespace then delete it?  I know that sounds fiddly and is something we would eventually want to unwind, but we probably need to do it if we want to get this done without complaints and bug reports about lost state from multiple clients/users.  @lbwexler your thoughts?

**Comment by lbwexler** [comment]
Yes, as you know I think that kind of runtime data cleanup/conversion fiddly, and normally not in favor of it,  but here we really would not have another choice, so I think we should carefully do it!

**Comment by lbwexler** [comment]
I could skip the hashing.  That's a false sense of security that just buys us troubleshooting difficulty.  The data will still be totally visible, and not appropriate for any real multi-user situation...

I think impersonation is the main reason to do this!

**Comment by TomTirapani** [comment]
Thanks for taking a look guys. I've added some migration logic, and removed the hashing.

File: svc/LocalStorageService.js [hunk 1]
```diff
@@ -55,6 +63,21 @@ export class LocalStorageService {
     //------------------
     getInstance() {
         throwIf(!this._supported, 'Local Storage is not supported');
-        return store.namespace(XH.appName);
+        return store.namespace(this.getNamespace());
+    }
+
+    getNamespace() {
+        return `${XH.appCode}.${XH.getUsername().toLowerCase()}`;
```
Comment (amcclain) [inline svc/LocalStorageService.js]: Technically we are requiring that HoistUser normalize usernames to lowercase:

https://github.com/exhi/hoist-core/blob/ff105dce5a563ba9ac27fb876a736449324012c8/src/main/groovy/io/xh/hoist/user/HoistUser.groovy#L22

Any reason to believe that mixed-case usernames are making it to the client?  I can understand the instinct to be defensive here and don't feel strongly about it, but that lowercase requirement is probably something we should take as a requirement / given throughout our client-side Hoist code.

File: svc/LocalStorageService.js [hunk 2]
```diff
@@ -55,6 +63,21 @@ export class LocalStorageService {
...
     }
+
+    migrateNamespace() {
+        const oldNamespace = store.namespace(XH.appName),
+            namespace = store.namespace(this.getNamespace());
+
+        if (oldNamespace.size() && !namespace.size()) {
+            namespace.setAll(oldNamespace.getAll());
+            oldNamespace.clear();
```
Comment (amcclain) [inline svc/LocalStorageService.js]: That certainly is convenient!  😄

**Comment by lbwexler** [comment]
Implemented ATM's suggestion and tweaked this a bit more -- just being a bit paranoid, and looking to minimize our exposure to it ever bringing down the app.

Feel free to merge!

**Pull request merged**